### PR TITLE
Add operator parameters in Adios2StMan for data compression

### PR DESCRIPTION
**Pull request by JasonRuonanWang**
@rtobar Please help review.

After experimenting various data compression techniques in ADIOS2 for a while, we want to finalize the API so the ASKAP DINGO project can incorporate this into their production workflow.

This PR adds a std::vector<std::map<std::string, std::string>> typed container of parameters into Adios2StMan::Adios2StMan(), to pass data compression parameters from user applications down to the ADIOS2 library. Each element in std::vector<std::map<std::string, std::string>> is a map containing necessary information to add an ADIOS2 operator (data compressor in this case). A parameter map must have the key *Operator* to specify the data compressor, and the key *Variable* to specify the CTDS column (ADIOS2 variable) that this operator is attaching to. 

So far, the Bzip2, Blosc lossless compressors, and the ZFP and SZ lossy compressors have been tested with this feature. For more information about all available compressors and parameters, please refer to the ADIOS2 docs https://adios2.readthedocs.io/en/latest/

An example code: https://github.com/JasonRuonanWang/SparseTableCopy/blob/main/tableCopy.cc

This PR also fixes a bug for parsing transport parameters.

### CODE REVIEW

**Comment by JasonRuonanWang** [comment]
@mpokorny This PR should make the compression feature of Adios2StMan a lot easier to configure than before. Feel free to have a try if you like. You would need to pull the latest ADIOS2 master branch too and recompile.

**Comment by gervandiepen** [comment]
I guess you know that you can attach a Record (key-value pairs) to a
column. That might be a nicer way to define the name of the compression
engine.

Op ma 30 aug. 2021 om 00:54 schreef Jason Wang ***@***.***>

**Comment by JasonRuonanWang** [comment]
> I guess you know that you can attach a Record (key-value pairs) to a column. That might be a nicer way to define the name of the compression engine. Op ma 30 aug. 2021 om 00:54 schreef Jason Wang ***@***.***>
> […](#)

I actually don't know how to use it...

I know @rtobar has added some Record parsing code for other engine and transport parameters. I can follow what he has done and add it for operator parameters, but I am not sure if I am able to read it before creating a column, if it's attached to a column itself. Because the ADIOS2 API currently needs this information while creating an ADIOS2 object, when it has not created any of the columns yet.

**Comment by JasonRuonanWang** [comment]
@gervandiepen @rtobar I have added Record for operator parameters.

File: tables/DataMan/Adios2StMan.cc [hunk 1]
```diff
@@ -144,23 +147,23 @@ rownr_t Adios2StMan::getNrRows()
...
+
 Adios2StMan::impl::impl(Adios2StMan &parent, MPI_Comm mpiComm)
     : parent(parent)
 {
     itsMpiComm = mpiComm;
-    std::string engineType;
-    adios2::Params engineParams;
-    std::vector<adios2::Params> transportParams;
-    Adios2StMan(mpiComm, engineType, engineParams, transportParams);
+    Adios2StMan(mpiComm, "", {}, {}, {});
```
Comment (rtobar) [inline tables/DataMan/Adios2StMan.cc:155]: The old version has the same problem, but I just saw it... should this be a delegating constructor in the member initialiser list, or am I misreading something? Otherwise we're just creating an `Adios2StMan` and quickly disposing of it, and the only thing that gets set is `itsMpiComm`. I guess this particular flavour of the constructor isn't used (much), otherwise we would have stepped into problems earlier.

In other words, I suspect this should be something like:

```c++
Adios2StMan::impl::impl(Adios2StMan &parent, MPI_Comm mpiComm)
  : Adios2StMan::impl(parent, mpiComm, "", {}, {}, {})
```

Alternatively we could rework the constructor list to have a single one with most of the arguments defaulted.

File: tables/DataMan/Adios2StManImpl.h [hunk 2]
```diff
@@ -101,6 +105,8 @@ class Adios2StMan::impl
     static constexpr const char *SPEC_FIELD_ENGINE_PARAMS = "ENGINEPARAMS";
     // The name of the specification field for the transport parameters
     static constexpr const char *SPEC_FIELD_TRANSPORT_PARAMS = "TRANSPORTPARAMS";
+    // The name of the specification field for the transport parameters
```
Comment (rtobar) [inline tables/DataMan/Adios2StManImpl.h:108]: ... for the "operator" parameters

**Review by rtobar — COMMENTED** [review]
@JasonRuonanWang I just had a look and the changes seem fine, just left a couple of comments.

I haven't tested these, and I probably won't have a chance to do it until I'm back at work next April. I trust you have though, so I'm happy with that. I'm sure whoever ends up using this in DINGO will also let us know if there's anything wrong.

I agree with you that if you need the per-column information before the columns are created (or at least written to, you could delay the creation of the files on disk until the very last moment) then you'll have to receive it at StMan construction time. I ran into a very similar issue in our [plasma storage manager](https://gitlab.com/ska-telescope/plasma-storage-manager), where I need to know (and store) certain column-specific information at the table metadata level. Having said that, there's always corners or the CDTS that I haven't explored, so it might be worth double-checking with @gervandiepen that we're doing things nicely to at least be aware of all possibilities and adjust the code in the future if necessary.

**Review by rtobar — COMMENTED** [review]
@JasonRuonanWang I just had a look and the changes seem fine, just left a couple of comments.

I haven't tested these, and I probably won't have a chance to do it until I'm back at work next April. I trust you have though, so I'm happy with that. I'm sure whoever ends up using this in DINGO will also let us know if there's anything wrong.

I agree with you that if you need the per-column information before the columns are created (or at least written to, you could delay the creation of the files on disk until the very last moment) then you'll have to receive it at StMan construction time. I ran into a very similar issue in our [plasma storage manager](https://gitlab.com/ska-telescope/plasma-storage-manager), where I need to know (and store) certain column-specific information at the table metadata level. Having said that, there's always corners or the CDTS that I haven't explored, so it might be worth double-checking with @gervandiepen that we're doing things nicely to at least be aware of all possibilities and adjust the code in the future if necessary.

File: tables/DataMan/Adios2StMan.cc (hunk 1 above)
Comment (JasonRuonanWang) [reply tables/DataMan/Adios2StMan.cc:155]: @rtobar I was thinking about having a single constructor too. I thought this was used somewhere so you haven't removed. If I remember correctly, this has nothing special besides that it can quickly instantiate an Adios2StMan with all default configurations without going through the logics in the other constructor. But those logics are not expensive anyway, so I guess I will just remove it.

File: tables/DataMan/Adios2StManImpl.h (hunk 2 above)
Comment (JasonRuonanWang) [reply tables/DataMan/Adios2StManImpl.h]: Fixed.

File: tables/DataMan/Adios2StMan.cc (hunk 1 above)
Comment (JasonRuonanWang) [reply tables/DataMan/Adios2StMan.cc]: This constructor is removed. Now there is only one constructor of Adios2StMan and Adios2StMan::impl, with default parameters in place.

**Comment by tammojan** [comment]
Hi @JasonRuonanWang , so sorry for missing this Pull Request! The changes look good to me; is it still ok to merge from your perspective?

**Comment by rtobar** [comment]
If it's any good, I'll finally be back at work next week, so I could test these changes if required.

**Comment by JasonRuonanWang** [comment]
@tammojan Thanks for taking care of this PR! Since it's been quite long time, I'd rebase and test, and see if everything is still good. Will let you know soon when it's ready to go. 

BTW, I pushed another commit which passes a XML file name to ADIOS2 for it to initialize operators with XML configurations. This is required by the ASKAP team recently, and it should be the final bit of work for this whole operator configuration thing.

**Comment by JasonRuonanWang** [comment]
@tammojan I have rebased this on the current casacore/master and re-tested. Everything turned to be good. So all good from my side. Please go ahead unless @pelahi or @rtobar have any objections.

**Comment by tammojan** [comment]
Thanks for the quick action @JasonRuonanWang ! I'll wait at least a week so @rtobar can have a look, and merge before we release a next version in a few weeks.

**Comment by gervandiepen** [comment]
Why using XML as a configuration file? Will it add a dependency on an XML parser? Casacore already has JSON parsing classes, so a JSON configuration file might be nicer.

**Comment by JasonRuonanWang** [comment]
> Why using XML as a configuration file? Will it add a dependency on an XML parser? Casacore already has JSON parsing classes, so a JSON configuration file might be nicer.

No, it won't add anything to casacore. It only passes a filename of a XML file to ADIOS2. ADIOS2 will then deal with the XML file.

**Comment by rtobar** [comment]
I'm back to work today so I'll try to test these changes today and will post my results here.

File: tables/DataMan/Adios2StMan.cc [hunk 3]
```diff
@@ -283,23 +339,45 @@ Record Adios2StMan::impl::dataManagerSpec() const
...
+            if (itVar == params.end()) {
+                continue;
+            }
+            auto itOper = params.find("Operator");
+            if (itOper == params.end()) {
+                continue;
+            }
+            operator_params_record.defineRecord(itVar->second, to_record(params));
+        }
+        record.defineRecord(SPEC_FIELD_TRANSPORT_PARAMS, operator_params_record);
```
Comment (rtobar) [inline tables/DataMan/Adios2StMan.cc:364]:
```suggestion
        record.defineRecord(SPEC_FIELD_OPERATOR_PARAMS, operator_params_record);
```

**Review by rtobar — COMMENTED** [review]
I just tried this after rebuilding my whole stack (compression libraries, ADIOS2, casacore, python-casacore). I ran my [small test program](https://gist.github.com/rtobar/700878c154272a3cdec9987d93ca8f52) which seems to be working fine:

```
$ python create_adios_table.py lala-bzip.tab bzip2 10000 3 3 10
Will write 6.87 MB of data into lala-bzip.tab
Table summary: /home/rtobar/lala-bzip.tab
Shape: 1 columns by 10000 rows
Info: {'type': '', 'subType': '', 'readme': ''}
Columns: ['DATA']

Structure of table /home/rtobar/lala-bzip.tab
------------------ 
10000 rows, 1 columns in little endian format (using 1 data managers)

 Adios2StMan file=table.f0  name=Adios2StMan
    TRANSPORTPARAMS=  DATA: {
    Accuracy: String "0.1"
    Operator: String "bzip2"
    Variable: String "DATA"
  }

  DATA double   shape=[10,3,3]

Table created and written in 3.09 s
Table size on disk: 0.01 MB
Compression ratio: 0.00


$ python create_adios_table.py lala-zfp.tab zfp 10000 3 3 10
Will write 6.87 MB of data into lala-zfp.tab
Table summary: /home/rtobar/lala-zfp.tab
Shape: 1 columns by 10000 rows
Info: {'type': '', 'subType': '', 'readme': ''}
Columns: ['DATA']

Structure of table /home/rtobar/lala-zfp.tab
------------------ 
10000 rows, 1 columns in little endian format (using 1 data managers)

 Adios2StMan file=table.f0  name=Adios2StMan
    TRANSPORTPARAMS=  DATA: {
    Accuracy: String "0.1"
    Operator: String "zfp"
    Variable: String "DATA"
  }

  DATA double   shape=[10,3,3]

Table created and written in 2.16 s
Table size on disk: 0.11 MB
Compression ratio: 0.02


$ python create_adios_table.py lala-blosc.tab blosc 10000 3 3 10
Will write 6.87 MB of data into lala-blosc.tab
[Wed Apr 13 15:22:59 2022] [ADIOS2 ERROR] <Operator> <CompressBlosc> <Operate> : ADIOS WARNING: Unknown parameter keyword 'accuracy' with value '0.1' passed to Blosc compression operator.
Table summary: /home/rtobar/lala-blosc.tab
Shape: 1 columns by 10000 rows
Info: {'type': '', 'subType': '', 'readme': ''}
Columns: ['DATA']

Structure of table /home/rtobar/lala-blosc.tab
------------------ 
10000 rows, 1 columns in little endian format (using 1 data managers)

 Adios2StMan file=table.f0  name=Adios2StMan
    TRANSPORTPARAMS=  DATA: {
    Accuracy: String "0.1"
    Operator: String "blosc"
    Variable: String "DATA"
  }

  DATA double   shape=[10,3,3]

Table created and written in 2.14 s
Table size on disk: 0.04 MB
Compression ratio: 0.01
```

I found however a small copy-paste error when generating the data manager description (note how in the table descriptions above it reads `TRANSPORTPARAMS` instead of `OPERATORPARAMS`), for which I left a suggestion.

File: tables/DataMan/Adios2StMan.cc (hunk 3 above)
Comment (JasonRuonanWang) [reply tables/DataMan/Adios2StMan.cc]: Fixed. Thanks for pointing out.

**Comment by tammojan** [comment]
Thanks for testing @rtobar, also good to have this example script! I hope I can remember where to find it.

And thanks for the improvements @JasonRuonanWang.
